client/files.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"net/url"
+	"os"
+	"strconv"
+	"time"
+)
+
+// ListFilesOptions holds the options for a call to ListFiles.
+type ListFilesOptions struct {
+	// Path is the absolute path of the file system entry to be listed.
+	Path string
+
+	// Pattern is the glob-like pattern string to filter results by. When
+	// present, only file system entries with names matching this pattern will
+	// be included in the call results.
+	Pattern string
+
+	// Itself, when set, will force directory entries not to be listed, but
+	// instead have their information returned as if they were regular files.
+	Itself bool
+}
+
+type FileInfo struct {
+	name    string
+	size    int64
+	mode    os.FileMode
+	modTime time.Time
+	path    string
+	userID  int
+	groupID int
+	user    string
+	group   string
+}
```

Comment:
It would probably be good to add a compile-time check that this implements `os.FileInfo`. In Go a common idiom to do that is adding the line (at the package level):

```go
var _ os.FileInfo = (*FileInfo)(nil)
```

Oh wait, that makes me realize: I think the receiver for all of these should be a pointer receiver, to avoid copying this relatively large struct (and to match the `[]*FileInfo` return value of `ListFiles`).